### ISSUE TITLE
Modifying Flows returned by Flow.copy modifies the original flow

### ISSUE THREAD

**Issue by jbweston** [PrefectHQ/prefect#3319]
## Description
If I modify (using `.replace`) a flow that was created by calling `.copy()` on an existing flow, the original flow is modified also.

## Expected Behavior
Modifying a *copy* of a flow should not modify the original flow.

## Reproduction

```python
@prefect.task
def f(x):
    return x

with prefect.Flow("foo") as original_flow:
    x = prefect.Parameter(name="x")
    y = f(x)
    
print(original_flow.constants)

copied_flow = original_flow.copy()
copied_flow.replace(x, 1)

print(original_flow.constants)
```

This prints
```
defaultdict(<class 'dict'>, {})
defaultdict(<class 'dict'>, {<Task: f>: {'x': 1}})
```

This means that calling `copied_flow.replace` modifies `original_flow`.

## Environment

```
{
  "config_overrides": {},
  "env_vars": [],
  "system_information": {
    "platform": "Linux-3.10.0-957.27.2.el7.x86_64-x86_64-with-glibc2.10",
    "prefect_backend": "cloud",
    "prefect_version": "0.13.6",
    "python_version": "3.8.5"
  }
}
```

**Comment by joshmeek** [comment]
Hi @jbweston I believe this functionality is expected and intentional where `flow.copy` is not intended to perform a deep copy. Is there a reason using something like `copy.deepcopy(your_flow)` will not suffice?

**Comment by jbweston** [comment]
@joshmeek there is no reason that I cannot use `copy.deepcopy`.

Perhaps the docstring of `Flow.copy` could be updated to stipulate that this is a shallow copy?

Also I am unsure how to interpret the term "shallow copy" in this case; as far as I can see the implementation of `Flow.copy` goes "1 level deeper" than one would expect of a shallow copy:
```python
        new = copy.copy(self)
        # create a new cache
        new._cache = dict()
        new.tasks = self.tasks.copy()
        new.edges = self.edges.copy()
        new.set_reference_tasks(self._reference_tasks)
```

It is also not clear to me why `replace` would not work as expected, given the above implementation of `Flow.copy`.

Naively I would think that `replace` merely modifies `self.tasks`, and given that a shallow copy of `self.tasks` is performed by `Flow.copy` then modifying this datastructure should still not affect the original.

**Comment by jbweston** [comment]
@joshmeek what should I do if I wanted to make a copy of the *flow* but *not* make deep copies of the Task instances contained therein?

**Comment by cicdw** [comment]
Hi @jbweston - this was actually a bug (we were not copying the `.constants` attribute when we should have been!).  PR to fix here: #3345 

Thanks for the bug report!
